internal/resource/resource.go

```diff
@@ -145,6 +133,115 @@ func (r *Resource) patchSetsDeletionTimestamp() bool {
 	return dt != ""
 }
 
+type SchemaGetter interface {
+	Get(ctx context.Context, gvk schema.GroupVersionKind) (typeref *smdschema.TypeRef, schem *smdschema.Schema, err error)
+}
+
+// Merge performs a three-way merge between the resource, it's old/previous Resource, and the current state.
+// Falls back to a non-structured three-way merge if the SchemaGetter returns a nil TypeRef.
+func (r *Resource) Merge(ctx context.Context, old *Resource, current *unstructured.Unstructured, sg SchemaGetter) (*unstructured.Unstructured, bool /* typed */, error) {
+	typeref, schem, err := sg.Get(ctx, r.GVK)
+	if err != nil {
+		return nil, false, fmt.Errorf("looking up schema: %w", err)
+	}
+
+	// Naive three-way merge for unknown types
+	if typeref == nil {
+		currentJS, err := current.MarshalJSON()
+		if err != nil {
+			return nil, false, fmt.Errorf("encoding current state: %w", err)
+		}
+
+		var prevJS []byte
+		if old != nil {
+			prevJS = []byte(old.Manifest.Manifest)
+		}
+
+		patch, err := jsonmergepatch.CreateThreeWayJSONMergePatch(prevJS, []byte(r.Manifest.Manifest), currentJS)
+		if err != nil {
+			return nil, false, fmt.Errorf("building merge patch: %w", err)
+		}
+		patchedJSON, err := jsonpatch.MergePatch(currentJS, patch)
+		if err != nil {
+			return nil, false, fmt.Errorf("applying merge patch: %w", err)
+		}
+
+		patched := &unstructured.Unstructured{}
+		err = patched.UnmarshalJSON(patchedJSON)
+		if err != nil {
+			return nil, false, fmt.Errorf("parsing patched resource: %w", err)
+		}
+
+		if equality.Semantic.DeepEqual(current, patched) || compareWithScheme(current, patched) {
+			return nil, false, nil
+		}
+		return patched, false, nil
+	}
+
+	// Convert to SMD values
+	currentVal := value.NewValueInterface(current.Object)
+	typedNew, err := typed.AsTyped(r.value, schem, *typeref)
+	if err != nil {
+		return nil, false, fmt.Errorf("converting new version to typed: %w", err)
+	}
+	typedCurrent, err := typed.AsTyped(currentVal, schem, *typeref)
+	if err != nil {
+		return nil, false, fmt.Errorf("converting current state to typed: %w", err)
+	}
+
+	// Merge properties that are set in the new state onto the current state
+	merged, err := typedCurrent.Merge(typedNew)
+	if err != nil {
+		return nil, false, fmt.Errorf("merging new state into current: %w", err)
+	}
+
+	// Prune properties that were present in the old state but not the new
+	if old != nil {
+		typedOld, err := typed.AsTyped(old.value, schem, *typeref)
+		if err != nil {
+			return nil, false, fmt.Errorf("converting old version to typed: %w", err)
+		}
+		toOld, err := typedOld.Compare(typedNew)
+		if err != nil {
+			return nil, false, fmt.Errorf("comparing new and old states: %w", err)
+		}
+		merged = merged.RemoveItems(toOld.Removed)
+	}
+
+	// Bail out if no changes are required
+	cmp, err := merged.Compare(typedCurrent)
+	if err == nil && cmp.IsSame() {
```

Comment:
Should be bail early if err != nil?

Reply:
I'm more inclined to fail open if we're for some reason unable to compare since an update applied multiple times is better than no update (in most cases). I'd like to think this code isn't reachable anyway but we all know how that goes